Add hsmKeyStore implementation

nklaassen opened the pull request:
This PR adds the `hsmKeyStore` implementation of the `KeyStore` inteferface, for interacting with PKCS#11 HSM private keys.

Basic tests are added for `SoftHSM2`, `YubiHSM2`, and `CloudHSM`.

`softhsm2` is added as a dependency to the docker buildbox for testing.

nklaassen committed:
replace GenerateSelfSignedCAWithPrivateKey with GenerateSelfSignedCAWithSigner

nklaassen committed:
Add KeyStore interface with rawKeyStore implementation

nklaassen committed:
add hsmKeyStore implementation

nklaassen committed:
switch to all 0s in test uuid

nklaassen committed:
hsm keystore should not fall back to raw keys

nklaassen committed:
Merge branch 'master' into nklaassen/hsm-2

nklaassen committed:
Merge branch 'nklaassen/hsm-2' into nklaassen/hsm-3

nklaassen committed:
Merge branch 'nklaassen/hsm-3' into nklaassen/hsm-4

tcsc commented on lib/auth/keystore/hsm.go:
```diff
@@ -0,0 +1,286 @@
+		if err != nil {
+			return nil, nil, err
+		}
+		signer, err = c.ctx.GenerateRSAKeyPairWithLabel(id[:], label, teleport.RSAKeySize)
```
So I'm assuming that `GenerateRSAKeyPairWithLabel()` will fail if there is an ID collision - is that true? Is it worth putting in some sort of test to verify this?

nklaassen replied:
That's what I was assuming as well, good catch because actually SoftHSM actually will *not* return any error for a duplicate ID (it does let us use the full 16 byte UUID though so we shouldn't get any collisions). YubiHSM does return an error, I haven't tried CloudHSM yet. I'm going to think a bit more about what to do for generating IDs because I don't really like what I have here

nklaassen replied:
I've updated this function to first check if the (possibly silently truncated) ID is already taken before generating the keypair

nklaassen committed:
add comment

nklaassen committed:
Update lib/auth/keystore/keystore.go

Co-authored-by: Brian Joerger <bjoerger@gravitational.com>

nklaassen committed:
use CA constructor in test

nklaassen committed:
remove redundant error returns

nklaassen committed:
Merge branch 'nklaassen/hsm-3' into nklaassen/hsm-4

nklaassen committed:
check for existing ID before generating keypair

nklaassen committed:
satisfy linter

xacrimon reviewed (approved):
Looks good overall but I can't comment/verify the correctness of HSM implementation details, variations and PKCS11 stuff

nklaassen committed:
Merge branch 'master' into nklaassen/hsm-4

nklaassen committed:
address review comments

- use doc.go instead of README
- add missing license header in hsm.go
- use trace.Wrap
- improve error messages
- return errors from findKeyPair
- factor out findUnusedID function

nklaassen commented on lib/auth/keystore/README.md:
```diff
@@ -0,0 +1,51 @@
+# Keystore
```
good point, did this in the latest comit

nklaassen commented on lib/auth/keystore/hsm.go:
```diff
@@ -0,0 +1,303 @@
+package keystore
```
fixed

nklaassen commented on lib/auth/keystore/hsm.go:
```diff
@@ -0,0 +1,303 @@
+		// some HSMs (like YubiHSM2) will return signer == nil && err == nil if
+		// no match is found
+		if err != nil || existingSigner == nil {
+			// failed to find an existing keypair, so this ID is unique
```
Yeah it turns out I was using `FindKeyPair` incorrectly, the returned signer is expected to be `nil` if it is not found, an error will only be returned if there is an unexpected error so I've updated this to propagate any error

nklaassen commented on lib/auth/keystore/hsm.go:
```diff
@@ -0,0 +1,303 @@
+	return nil, trace.BadParameter("unrecognized key type %s", keyType.String())
+}
+
+func (c *hsmKeyStore) selectTLSKeyPair(ca types.CertAuthority) (*types.TLSKeyPair, error) {
```
I don't really see a use for it outside of the `KeyStore`. I've tried to abstract away the "do I have an HSM or not" concept so that most code outside of this package does not need to care and we don't have conditionals everywhere. If it's needed elsewhere this can easily be refactored.

nklaassen commented on lib/auth/keystore/hsm.go:
```diff
@@ -0,0 +1,303 @@
+	// 2(rotation phases) * (4(SSH and TLS for User and Host CA) + 1(JWT CA))
+	maxIterations := 16
+	iterations := 0
+	for ; iterations < maxIterations; iterations++ {
```
yeah good call

nklaassen committed:
Merge branch 'master' into nklaassen/hsm-4

nklaassen committed:
import api v7
